File: src/main/java/org/cimsbioko/activity/FieldWorkerActivity.java
package org.cimsbioko.activity;

import android.app.SearchManager;
import android.app.SearchableInfo;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.widget.Toolbar;
import androidx.core.view.MenuItemCompat;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.SearchView;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.View.OnClickListener;
import android.widget.LinearLayout;
import android.widget.RelativeLayout;

import org.cimsbioko.R;
import org.cimsbioko.fragment.navigate.FormListFragment;
import org.cimsbioko.model.core.FieldWorker;
import org.cimsbioko.navconfig.NavigatorModule;
import org.cimsbioko.utilities.ConfigUtils;

import java.util.Iterator;

import static org.cimsbioko.search.Utils.isSearchEnabled;
import static org.cimsbioko.utilities.LayoutUtils.makeTextWithPayload;
import static org.cimsbioko.utilities.LoginUtils.getLogin;
import static org.cimsbioko.utilities.FormsHelper.getAllUnsentFormInstances;

public class FieldWorkerActivity extends AppCompatActivity implements OnClickListener {

    public static final String ACTIVITY_MODULE_EXTRA = "ACTIVITY_MODULE_EXTRA";

    private FormListFragment formListFragment;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.fieldworker_activity);
        setTitle(getText(R.string.field_worker_home));

        Toolbar toolbar = findViewById(R.id.fieldworker_toolbar);
        setSupportActionBar(toolbar);
        ActionBar actionBar = getSupportActionBar();
        if (actionBar != null) {
            actionBar.setDisplayHomeAsUpEnabled(true);
            actionBar.setHomeButtonEnabled(true);
        }

        LinearLayout activitiesLayout = findViewById(R.id.portal_middle_column);
        Iterator<NavigatorModule> moduleIter = ConfigUtils.getActiveModules(this).iterator();
        while (moduleIter.hasNext()) {
            NavigatorModule module = moduleIter.next();
            RelativeLayout layout = makeTextWithPayload(this, module.getLaunchLabel(), module.getLaunchDescription(),
                    module.getName(), this, activitiesLayout, R.drawable.data_selector, null, null, true);
            LinearLayout.LayoutParams params = (LinearLayout.LayoutParams) layout.getLayoutParams();
            if (moduleIter.hasNext()) {
                params.setMargins(0, 0, 0, getResources().getDimensionPixelSize(R.dimen.module_button_spacing));
            }
        }

        formListFragment = (FormListFragment) getSupportFragmentManager().findFragmentById(R.id.portal_form_list);
        formListFragment.setHeaderText(R.string.unsent_forms);
        formListFragment.setFindEnabled(true);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {

        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.fieldworker_menu, menu);

        MenuItem searchMenuItem = menu.findItem(R.id.field_worker_search);
        boolean searchEnabled = isSearchEnabled(this);
        if (searchEnabled) {
            SearchManager searchManager = (SearchManager) getSystemService(Context.SEARCH_SERVICE);
            SearchView searchView = (SearchView) MenuItemCompat.getActionView(menu.findItem(R.id.field_worker_search));
            SearchableInfo searchInfo = searchManager.getSearchableInfo(new ComponentName(this, SearchableActivity.class));
            searchView.setSearchableInfo(searchInfo);
        }
        searchMenuItem.setVisible(searchEnabled);

        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.logout_menu_button:
                getLogin(FieldWorker.class).logout(this, true);
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    @Override
    public void onClick(View v) {
        Intent intent = new Intent(this, HierarchyNavigatorActivity.class);
        String activityName = (String) v.getTag();
        intent.putExtra(ACTIVITY_MODULE_EXTRA, activityName);
        startActivity(intent);
    }

    @Override
    protected void onResume() {
        super.onResume();
        populateFormInstanceListView();
    }

    private void populateFormInstanceListView() {
        formListFragment.populate(getAllUnsentFormInstances());
    }
}

File: src/main/java/org/cimsbioko/activity/FieldWorkerActivity.kt
package org.cimsbioko.activity

import android.app.SearchManager
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.LinearLayout
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.SearchView
import androidx.appcompat.widget.Toolbar
import androidx.core.view.MenuItemCompat
import org.cimsbioko.R
import org.cimsbioko.fragment.navigate.FormListFragment
import org.cimsbioko.search.Utils.isSearchEnabled
import org.cimsbioko.utilities.ConfigUtils.getActiveModules
import org.cimsbioko.utilities.FormsHelper.allUnsentFormInstances
import org.cimsbioko.utilities.LayoutUtils.makeTextWithPayload
import org.cimsbioko.utilities.LoginUtils.login

class FieldWorkerActivity : AppCompatActivity(), View.OnClickListener {

    private lateinit var formListFragment: FormListFragment

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.fieldworker_activity)
        title = getText(R.string.field_worker_home)
        findViewById<Toolbar>(R.id.fieldworker_toolbar).also { setSupportActionBar(it) }
        supportActionBar?.apply {
            setDisplayHomeAsUpEnabled(true)
            setHomeButtonEnabled(true)
        }
        val activitiesLayout = findViewById<LinearLayout>(R.id.portal_middle_column)
        getActiveModules(this).also { modules ->
            val lastIndex = modules.indices.last
            for ((index, module) in modules.withIndex()) {
                makeTextWithPayload(this, module.launchLabel, module.launchDescription, module.name, this,
                        activitiesLayout, R.drawable.data_selector, null, null, true)
                        .takeIf { index != lastIndex }
                        ?.let { it.layoutParams as LinearLayout.LayoutParams }
                        ?.setMargins(0, 0, 0, resources.getDimensionPixelSize(R.dimen.module_button_spacing))
            }
        }
        formListFragment = (supportFragmentManager.findFragmentById(R.id.portal_form_list) as FormListFragment).apply {
            setHeaderText(R.string.unsent_forms)
            setFindEnabled(true)
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.fieldworker_menu, menu)
        menu.findItem(R.id.field_worker_search).isVisible = isSearchEnabled(this).also { enabled ->
            if (enabled) {
                val searchManager = getSystemService(Context.SEARCH_SERVICE) as SearchManager
                val searchView = MenuItemCompat.getActionView(menu.findItem(R.id.field_worker_search)) as SearchView
                val searchInfo = searchManager.getSearchableInfo(ComponentName(this, SearchableActivity::class.java))
                searchView.setSearchableInfo(searchInfo)
            }
        }
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == R.id.logout_menu_button) {
            login.logout(this, true)
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onClick(v: View) {
        startActivity(Intent(this, HierarchyNavigatorActivity::class.java).apply {
            putExtra(ACTIVITY_MODULE_EXTRA, v.tag as String)
        })
    }

    override fun onResume() {
        super.onResume()
        populateFormInstanceListView()
    }

    private fun populateFormInstanceListView() {
        formListFragment.populate(allUnsentFormInstances)
    }

    companion object {
        const val ACTIVITY_MODULE_EXTRA = "ACTIVITY_MODULE_EXTRA"
    }
}